apollo-router/src/plugins/authentication.rs

```diff
@@ -0,0 +1,944 @@
+                            tracing::info!(
+                                monotonic_counter.apollo_authentication_success_count = 1u64,
+                                kind = %AUTHENTICATION_KIND
+                            );
+                            Ok(ControlFlow::Continue(request))
+                        }
+                        None => {
+                            // We can't find this "kid". We will observe the COOLDOWN, if one is
+                            // set, to minimise the impact of DOS attacks via this vector.
+                            //
+                            // If there is no COOLDOWN, we'll trigger a cache update and set a
+                            // COOLDOWN (via the true flag to err_cleanup(true).
+                            if COOLDOWN.load(Ordering::SeqCst) {
+                                // This is a metric and will not appear in the logs
+                                tracing::info!(
+                                    monotonic_counter.apollo_authentication_cooldown_count = 1u64,
+                                    kind = %AUTHENTICATION_KIND
+                                );
+                                let response = router::Response::error_builder()
+                                    .error(
+                                        graphql::Error::builder()
+                                            .message(
+                                                "Could not retrieve JWKS set: router cooling down",
+                                            )
+                                            .extension_code("AUTH_ERROR")
+                                            .build(),
+                                    )
+                                    .header(
+                                        http::header::RETRY_AFTER,
+                                        my_config
+                                            .cooldown
+                                            .unwrap_or(DEFAULT_AUTHENTICATION_COOLDOWN)
+                                            .as_secs()
+                                            .to_string(),
+                                    )
+                                    .status_code(StatusCode::SERVICE_UNAVAILABLE)
+                                    .context(request.context)
+                                    .build()?;
+                                Ok(ControlFlow::Break(response))
+                            } else {
+                                err_cleanup(true); // c.
+                                failure_message(
+                                    request.context,
+                                    format!("Could not find kid: '{kid}' in JWKS set"),
+                                    StatusCode::UNAUTHORIZED,
+                                )
+                            }
+                        }
+                    }
+                }
+            })
+            .buffer(20_000)
```

Comment:
so, I have to ask, why is this buffered? :smile:

Reply:
If I don't buffer the service, then I get:

```
error[E0277]: the trait bound `tower::util::BoxService<services::router::Request, services::router::Response, Box<(dyn StdError + std::marker::Send + std::marker::Sync + 'static)>>: Clone` is not satisfied
   --> apollo-router/src/plugins/authentication.rs:501:22
    |
501 |             .service(service)
    |              ------- ^^^^^^^ the trait `Clone` is not implemented for `tower::util::BoxService<services::router::Request, services::router::Response, Box<(dyn StdError + std::marker::Send + std::marker::Sync + 'static)>>`
    |              |
    |              required by a bound introduced by this call
    |
    = help: the trait `tower::Layer<S>` is implemented for `Stack<Inner, Outer>`
<plus about another couple of hundred lines of errror...>
```

This all stems from `checkpoint_async` which requires the service to be `Clone`. I presume you are really asking "why not make it a layer rather than a plugin?".

Reply:
this can land with the buffer in there, we'll see if we need to remove it later